### PR TITLE
Fix Re-indexing Dependency Injection

### PR DESCRIPTION
**Pull request by wsugarman**
## Description
- Fix dependency injection service container to include missing services like the SQL index stores, JSON.NET serializers, and RecycleableMemoryStreamManager
- Normalize serialization between web server and client to use System.Text.Json over JSON.NET in preparation for fo-dicom 5 and .NET 6

## Related issues
N/A

## Testing
Test locally

### CODE REVIEW

File: src/Microsoft.Health.Dicom.Functions.Client.UnitTests/DicomAzureFunctionsHttpClientTests.cs [hunk 1]
```diff
@@ -35,19 +38,30 @@ public class DicomAzureFunctionsHttpClientTests
                 }
             });
 
+        public DicomAzureFunctionsHttpClientTests()
+        {
+            var options = new JsonSerializerOptions();
+            options.Converters.Add(new JsonStringEnumConverter());
+            _jsonSerializerOptions = Options.Create(options);
+        }
```
Comment (yuedai0531) [inline src/Microsoft.Health.Dicom.Functions.Client.UnitTests/DicomAzureFunctionsHttpClientTests.cs:46]: is it normal that this method is not used anywhere?

File: src/Microsoft.Health.Dicom.Functions.Client.UnitTests/DicomAzureFunctionsHttpClientTests.cs [hunk 2]
```diff
@@ -8,23 +8,26 @@
...
 using Xunit;
 
 namespace Microsoft.Health.Dicom.Functions.Client.UnitTests
 {
     public class DicomAzureFunctionsHttpClientTests
     {
-        private static readonly IOptions<FunctionsClientOptions> DefaultConfig = Options.Create(
+        private readonly IOptions<JsonSerializerOptions> _jsonSerializerOptions;
+
+        private static readonly IOptions<FunctionsClientOptions> DefaultOptions = Options.Create(
```
Comment (yuedai0531) [inline src/Microsoft.Health.Dicom.Functions.Client.UnitTests/DicomAzureFunctionsHttpClientTests.cs:30]: going forward are we changing all Config to be Options?

Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Functions.Client.UnitTests/DicomAzureFunctionsHttpClientTests.cs:30]: I think generally we should create a distinction between the concept of "options," a collection of settings, and a "config," a collection of settings sourced from a particular storage medium such as a file. In .NET Core, it appears these settings are often called "options" and are loaded into the DI container via the `IOptions<T>`, `IOptionsSnapshot<T>`, and `IOptionsMonitor<T>`. It also just so happens that the data in an `IOptions<T>` instance is sourced from an `IConfiguration` object, but it is not a requirement.

File: src/Microsoft.Health.Dicom.Functions.Client.UnitTests/DicomAzureFunctionsHttpClientTests.cs [hunk 3]
```diff
@@ -110,16 +124,16 @@ public async Task GivenSuccessfulResponse_WhenGettingStatus_ThenReturnStatus()
 @$"
 {{
   ""OperationId"": ""{id}"",
-  ""Type"": ""Reindex"",
-  ""CreatedTime"": ""{createdDateTime}"",
-  ""LastUpdatedTime"": ""{createdDateTime.AddMinutes(15)}"",
-  ""Status"": ""Running""
+  ""Type"": ""{OperationType.Reindex}"",
+  ""CreatedTime"": ""{createdDateTime:O}"",
+  ""LastUpdatedTime"": ""{createdDateTime.AddMinutes(15):O}"",
```
Comment (yuedai0531) [inline src/Microsoft.Health.Dicom.Functions.Client.UnitTests/DicomAzureFunctionsHttpClientTests.cs:129]: what is this for? `:O`

Comment (yuedai0531) [reply src/Microsoft.Health.Dicom.Functions.Client.UnitTests/DicomAzureFunctionsHttpClientTests.cs:129]: also is it normal to indent json like this?

File: src/Microsoft.Health.Dicom.Functions.Client.UnitTests/DicomAzureFunctionsHttpClientTests.cs (hunk 2 above)
Comment (yuedai0531) [reply src/Microsoft.Health.Dicom.Functions.Client.UnitTests/DicomAzureFunctionsHttpClientTests.cs:30]: should we rename this then?
image

File: src/Microsoft.Health.Dicom.Functions.Client.UnitTests/DicomAzureFunctionsHttpClientTests.cs (hunk 3 above)
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Functions.Client.UnitTests/DicomAzureFunctionsHttpClientTests.cs:129]: That is the [format specifier used by `DateTime`](https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-date-and-time-format-strings#table-of-format-specifiers) to indicate "Round Trip." .NET string interpolation uses the [following syntax](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/interpolated#structure-of-an-interpolated-string):

```
{<interpolationExpression>[,<alignment>][:<formatString>]}
```

File: src/Microsoft.Health.Dicom.Functions.Client.UnitTests/DicomAzureFunctionsHttpClientTests.cs (hunk 2 above)
Comment (yuedai0531) [reply src/Microsoft.Health.Dicom.Functions.Client.UnitTests/DicomAzureFunctionsHttpClientTests.cs:30]: oh oops nevermind

Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Functions.Client.UnitTests/DicomAzureFunctionsHttpClientTests.cs:30]: Yeah, I think that makes sense

File: src/Microsoft.Health.Dicom.Functions/Indexing/ReindexDurableFunction.Trigger.cs [hunk 4]
```diff
@@ -60,17 +57,18 @@ public partial class ReindexDurableFunction
...
+                    throw new JsonException("Expected at least 1 element in array.");
+                }
             }
-
-            if (extendedQueryTagKeys == null || extendedQueryTagKeys.Count == 0)
+            catch (JsonException e)
             {
+                logger.LogError(e, "Cannot deserialize extended query tag keys.");
                 return new HttpResponseMessage { StatusCode = HttpStatusCode.BadRequest };
             }
```
Comment (yuedai0531) [inline src/Microsoft.Health.Dicom.Functions/Indexing/ReindexDurableFunction.Trigger.cs:73]: https://media.giphy.com/media/mGK1g88HZRa2FlKGbz/giphy.gif

File: src/Microsoft.Health.Dicom.Functions/Management/DurableClientProxy.cs [hunk 5]
```diff
@@ -69,16 +80,20 @@ public static class DurableClientProxyFunctions
 
             if (string.IsNullOrWhiteSpace(instanceId))
             {
-                return new BadRequestResult();
+                return new HttpResponseMessage { StatusCode = HttpStatusCode.BadRequest };
```
Comment (yuedai0531) [inline src/Microsoft.Health.Dicom.Functions/Management/DurableClientProxy.cs:83]: what is the reason for the change here?

Comment (yuedai0531) [reply src/Microsoft.Health.Dicom.Functions/Management/DurableClientProxy.cs:83]: I think the majority of our controllers return an IActionResult? for consistency shouldn't we keep it like that?

**Review by yuedai0531 — APPROVED** [review]
j'ai fait mon possible :)

File: src/Microsoft.Health.Dicom.Functions/Management/DurableClientProxy.cs (hunk 5 above)
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Functions/Management/DurableClientProxy.cs:83]: So ASP.NET Core relies on middleware to perform serialization. So typically APIs return `IActionResult` which typically encapsulate an HTTP Status Code and optionally some payload. Middleware then takes that payload and serializes it into the proper content type, like JSON or XML. However, Azure Functions are special. It seems they've custom-built a bunch of it including the DI service provider. So unfortunately while ASP.NET Core 3.1 uses the new `System.Text.Json` lib for serialization that is part of the BCL, Azure Functions relies on Json.NET (aka `Newtonsoft.Json`). Furthermore, you can seem to swap it out, so I am forced to serialize it myself and directly return `HttpResponseMessage` with the serialized payload.

File: src/Microsoft.Health.Dicom.Functions/Indexing/ReindexDurableFunction.Trigger.cs (hunk 4 above)
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Functions/Indexing/ReindexDurableFunction.Trigger.cs:73]: lol I didn't want two different conditions to return `BadRequest`, so the deserialization may throw the exception... or I may do it directly 😛
